File: app/Models/Variavel.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use App\Traits\CalculaRangosVariavel;

class Variavel extends Model
{
    use HasFactory, CalculaRangosVariavel;

    protected $table = 'variaveis';

    protected $fillable = [
        'nome',
        'descricao',
        'tag',
        'B',
        'M',
        'A',
        'baixa',
        'moderada',
        'alta',

        'r_baixa',
        'r_moderada',
        'r_alta',

        'd_baixa',
        'd_moderada',
        'd_alta',
        
        'formulario_id',
    ];

    protected $casts = [
        'B' => 'integer',
        'M' => 'integer',
        'A' => 'integer',
    ];

    public function formulario()
    {
        return $this->belongsTo(Formulario::class, 'formulario_id');
    }

    public function perguntas()
    {
        return $this->belongsToMany(Pergunta::class, 'pergunta_variavel');
    }
    
    /**
     * Boot del modelo - actualiza rangos automáticamente cuando se actualizan las preguntas
     */
    protected static function boot()
    {
        parent::boot();
        
        // Actualizar rangos después de guardar, pero solo si cambió algo relevante
        static::saved(function ($variavel) {
            // Evitar bucle infinito: solo actualizar si los rangos necesitan cambio
            $variavel->load('perguntas', 'formulario');
            $totalPerguntas = $variavel->perguntas()->count();
            $scoreFim = $variavel->formulario->score_fim ?? 6;
            $rangos = static::calcularRangosGenerales($totalPerguntas, $scoreFim);
            
            // Solo actualizar si los valores son diferentes
            if ($variavel->B != $rangos['B'] || $variavel->M != $rangos['M'] || $variavel->A != $rangos['A']) {
                // Usar update sin eventos para evitar bucle
                static::withoutEvents(function() use ($variavel, $rangos) {
                    $variavel->update([
                        'B' => $rangos['B'],
                        'M' => $rangos['M'],
                        'A' => $rangos['A']
                    ]);
                });
            }
        });
    }
}

File: migrations/backend-api/app/schemas/variavel.py
"""
Schemas Pydantic para Variavel
"""
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

class VariavelBase(BaseModel):
    formulario_id: int
    nome: str
    descricao: Optional[str] = None
    tag: str
    B: int
    M: int
    A: Optional[int] = None
    baixa: Optional[str] = None
    moderada: Optional[str] = None
    alta: Optional[str] = None
    r_baixa: Optional[str] = None
    r_moderada: Optional[str] = None
    r_alta: Optional[str] = None
    d_baixa: Optional[str] = None
    d_moderada: Optional[str] = None
    d_alta: Optional[str] = None

class VariavelCreate(VariavelBase):
    pass

class VariavelUpdate(BaseModel):
    nome: Optional[str] = None
    descricao: Optional[str] = None
    tag: Optional[str] = None
    B: Optional[int] = None
    M: Optional[int] = None
    A: Optional[int] = None
    baixa: Optional[str] = None
    moderada: Optional[str] = None
    alta: Optional[str] = None
    r_baixa: Optional[str] = None
    r_moderada: Optional[str] = None
    r_alta: Optional[str] = None
    d_baixa: Optional[str] = None
    d_moderada: Optional[str] = None
    d_alta: Optional[str] = None

class VariavelResponse(VariavelBase):
    id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
